rpc_auth.go

```diff
@@ -0,0 +1,87 @@
+// grep -R -B 3 '"Pin"' | grep -C 1 '"Cluster"'.
+// This does not cover globalPinInfo*(...) broadcasts nor redirects to leader
+// in Raft.
+
+// DefaultRPCPolicy associates all rpc endpoints offered by cluster peers to an
+// endpoint type. See rpcutil/policygen.go as a quick way to generate this
+// without missing any endpoint.
+var DefaultRPCPolicy = map[string]RPCEndpointType{
+	// Cluster methods
+	"Cluster.BlockAllocate":      RPCClosed,
+	"Cluster.ConnectGraph":       RPCClosed,
+	"Cluster.ID":                 RPCOpen,
+	"Cluster.Join":               RPCClosed,
+	"Cluster.PeerAdd":            RPCOpen, // Used by Join()
+	"Cluster.PeerRemove":         RPCTrusted,
+	"Cluster.Peers":              RPCTrusted, // Used by ConnectGraph
+	"Cluster.Pin":                RPCClosed,
+	"Cluster.PinGet":             RPCClosed,
+	"Cluster.PinPath":            RPCClosed,
+	"Cluster.Pins":               RPCClosed, // stateless, proxy, api
+	"Cluster.Recover":            RPCClosed,
+	"Cluster.RecoverAllLocal":    RPCClosed,
+	"Cluster.RecoverLocal":       RPCClosed,
+	"Cluster.SendInformerMetric": RPCClosed, // Local use in ipfshttp
+	"Cluster.Status":             RPCClosed,
+	"Cluster.StatusAll":          RPCClosed,
+	"Cluster.StatusAllLocal":     RPCClosed,
+	"Cluster.StatusLocal":        RPCClosed,
+	"Cluster.Sync":               RPCClosed,
+	"Cluster.SyncAll":            RPCClosed,
+	"Cluster.SyncAllLocal":       RPCTrusted, // Bcast from SyncAll()
+	"Cluster.SyncLocal":          RPCTrusted, // Bcast from Sync()
+	"Cluster.Unpin":              RPCClosed,
+	"Cluster.UnpinPath":          RPCClosed,
+	"Cluster.Version":            RPCOpen,
+
+	// PinTracker methods
+	"PinTracker.Recover":    RPCTrusted, // Bcast from Cluster.Recover()
+	"PinTracker.RecoverAll": RPCClosed,  // Bcast unimplemented
+	"PinTracker.Status":     RPCTrusted, // Bcast from Cluster.Status()
+	"PinTracker.StatusAll":  RPCTrusted, // Bcast from Cluster.StatusAll()
+	"PinTracker.Track":      RPCClosed,
+	"PinTracker.Untrack":    RPCClosed,
+
+	// IPFSConnector methods
+	"IPFSConnector.BlockGet":      RPCClosed,
+	"IPFSConnector.BlockPut":      RPCTrusted, // Add() allocations
+	"IPFSConnector.ConfigKey":     RPCClosed,
+	"IPFSConnector.ConnectSwarms": RPCClosed,
+	"IPFSConnector.Pin":           RPCClosed,
+	"IPFSConnector.PinLs":         RPCClosed,
+	"IPFSConnector.PinLsCid":      RPCClosed,
+	"IPFSConnector.RepoStat":      RPCTrusted, // ipfsproxy: repo/stat broadcast
+	"IPFSConnector.Resolve":       RPCClosed,
+	"IPFSConnector.SwarmPeers":    RPCTrusted, // connectGraph.
+	"IPFSConnector.Unpin":         RPCClosed,
+
+	// Consensus methods
+	"Consensus.AddPeer":  RPCTrusted, // Raft redirect to leader
+	"Consensus.LogPin":   RPCTrusted, // Raft redirect to leader
+	"Consensus.LogUnpin": RPCTrusted, // Raft redirect to leader
+	"Consensus.Peers":    RPCClosed,
+	"Consensus.RmPeer":   RPCTrusted, // Raft redirect to leader
```

Comment:
In case of crdts, all consensus method could be `RPCClosed`, right?
This makes me wonder if we should have different default rpc policies for crdt and raft.

Reply:
We make the assumption trusted peers are trusted and therefore can call those methods (which they call indirectly anyway by doing pubsub broadcasts). Also add/rm peers are noops.